File: python/ex2/costFunction.py
import numpy as np
from sigmoid import sigmoid
# 计算代价和梯度
def cost_Function_Reg(X,Y,theta,lmd):
	m = X.shape[0]
	Z = np.dot(X,theta)
	g = sigmoid(Z)

	cost = - (Y.T).dot(np.log(g)) - ((1-Y).T).dot(np.log(1-g))
	cost = cost /(m) + lmd * (theta.T).dot(theta) / (2*m)
	# g.shape = [m,]
	# Y.shape = [m,1]
	# 两者相减得到的是[m,m]
	grad = (X.T).dot(g-Y.reshape(Y.size))/ m
	grad[0] = grad[0]
	grad[1:] = grad[1:] + (lmd * theta[1:])/m

	return cost,grad

File: matlab/ex2/costFunction.m
function [ cost,grad ] = costFunction( theta,X,Y )
%% 函数功能：根据输入的参数theta，数据的特征X，数据的输出Y，计算出此时的代价和梯度
%需要注意的是，logistic的代价函数不是最小均方差，而是似然函数
[m,~]=size(X);
z = sigmoid(X * theta);

cost = ( -Y.' * log(z) - (1 - Y.') * log(1 - z)) / m;

grad = - (Y.' - (z).') * X / m;

end

